File: client_2finance/cashback.go
package client_2finance


import (
	"time"
	"gitlab.com/2finance/2finance-network/blockchain/contract/cashbackV1"

	"gitlab.com/2finance/2finance-network/blockchain/encryption/keys"
	"fmt"
	"gitlab.com/2finance/2finance-network/blockchain/types"
)


// AddCashBack deploys a new cashback program (to = DEPLOY address).
func (c *networkClient) AddCashback(
	address string,
	owner string,
	tokenAddress string,
	programType string,  // "fixed-percentage" | "variable-percentage"
	percentage string,   // string-encoded number
	startAt time.Time,
	expiredAt time.Time,
	paused bool,
) (types.ContractOutput, error) {

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}

	if owner == "" {
		return types.ContractOutput{}, fmt.Errorf("owner not set")
	}
	if err := keys.ValidateEDDSAPublicKey(owner); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid owner address: %w", err)
	}
	if tokenAddress == "" {
		return types.ContractOutput{}, fmt.Errorf("token address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(tokenAddress); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid token address: %w", err)
	}
	if programType != "fixed-percentage" && programType != "variable-percentage" {
		return types.ContractOutput{}, fmt.Errorf("invalid program_type: %s", programType)
	}
	if percentage == "" {
		return types.ContractOutput{}, fmt.Errorf("percentage not set")
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_ADD_CASHBACK
	data := map[string]interface{}{
		"address":       address,
		"owner":         owner,
		"token_address": tokenAddress,
		"program_type":  programType,
		"percentage":    percentage,
		"start_at":      startAt,
		"expired_at":    expiredAt,
		"paused":        paused,
	}

	cashback, err := c.SignAndSendTransaction(from, to, contractVersion, method, data)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("failed to add cashback: %w", err)
	}
	return cashback, nil
}

// UpdateCashback updates an existing cashback program (to = program address). OnlyOwner.
func (c *networkClient) UpdateCashback(
	address string,
	tokenAddress string,
	programType string, // "fixed-percentage" | "variable-percentage"
	percentage string,
	startAt time.Time,
	expiredAt time.Time,
) (types.ContractOutput, error) {

	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if tokenAddress != "" {
		if err := keys.ValidateEDDSAPublicKey(tokenAddress); err != nil {
			return types.ContractOutput{}, fmt.Errorf("invalid token address: %w", err)
		}
	}
	if programType != "fixed-percentage" && programType != "variable-percentage" {
		return types.ContractOutput{}, fmt.Errorf("invalid program_type: %s", programType)
	}
	if percentage == "" {
		return types.ContractOutput{}, fmt.Errorf("percentage not set")
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_UPDATE_CASHBACK

	data := map[string]interface{}{
		"address":       address,
		"token_address": tokenAddress, // optional: include if your handler supports it
		"program_type":  programType,
		"percentage":    percentage,
		"start_at":      startAt,
		"expired_at":    expiredAt,
	}

	return c.SignAndSendTransaction(from, to, contractVersion, method, data)
}

// PauseCashBack pauses a cashback program. OnlyOwner.
func (c *networkClient) PauseCashback(address string, pause bool) (types.ContractOutput, error) {
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if !pause {
		return types.ContractOutput{}, fmt.Errorf("pause must be true: Pause: %t", pause)
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_PAUSE_CASHBACK

	data := map[string]interface{}{
		"address": address,
		"paused":  pause,
	}

	return c.SignAndSendTransaction(from, to, contractVersion, method, data)
}

// UnpauseCashback unpauses a cashback program. OnlyOwner.
func (c *networkClient) UnpauseCashback(address string, pause bool) (types.ContractOutput, error) {
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if pause {
		return types.ContractOutput{}, fmt.Errorf("pause must be false: Pause: %t", pause)
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_UNPAUSE_CASHBACK

	data := map[string]interface{}{
		"address": address,
		"paused":  pause,
	}

	return c.SignAndSendTransaction(from, to, contractVersion, method, data)
}

// DepositCashBack funds the cashback pool (token inferred from state).
func (c *networkClient) DepositCashbackFunds(address, tokenAddress, amount string) (types.ContractOutput, error) {
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if amount == "" {
		return types.ContractOutput{}, fmt.Errorf("amount not set")
	}

	if tokenAddress == "" {
		return types.ContractOutput{}, fmt.Errorf("token address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(tokenAddress); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid token address: %w", err)
	}
	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_DEPOSIT_CASHBACK

	data := map[string]interface{}{
		"address": address,
		"token_address": tokenAddress, // token address inferred from state
		"amount":  amount,
	}

	contractOutput, err := c.SignAndSendTransaction(from, to, contractVersion, method, data)
	if err != nil {
		return types.ContractOutput{}, fmt.Errorf("failed to deposit cashback: %w", err)
	}
	return contractOutput, nil
}

// WithdrawCashback withdraws funds from the cashback pool. OnlyOwner.
func (c *networkClient) WithdrawCashbackFunds(address, tokenAddress, amount string) (types.ContractOutput, error) {
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if amount == "" {
		return types.ContractOutput{}, fmt.Errorf("amount not set")
	}

	if tokenAddress == "" {
		return types.ContractOutput{}, fmt.Errorf("token address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(tokenAddress); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid token address: %w", err)
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_WITHDRAW_CASHBACK

	data := map[string]interface{}{
		"address": address,
		"amount":  amount,
		"token_address": tokenAddress, // token address inferred from state
	}

	return c.SignAndSendTransaction(from, to, contractVersion, method, data)
}

// GetCashBack reads a single cashback state.
func (c *networkClient) GetCashback(address string) (types.ContractOutput, error) {
	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("cashback address must be set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid cashback address: %w", err)
	}

	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_GET_CASHBACK
	data := map[string]interface{}{
		"address": address,
	}

	return c.GetState(contractVersion, method, data)
}

// ListCashBack queries cashback programs with filters + pagination.
func (c *networkClient) ListCashbacks(
	owner string,
	tokenAddress string,
	programType string,
	paused bool,
	page int,
	limit int,
	ascending bool,
) (types.ContractOutput, error) {
	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	if owner != "" {
		if err := keys.ValidateEDDSAPublicKey(owner); err != nil {
			return types.ContractOutput{}, fmt.Errorf("invalid owner address: %w", err)
		}
	}
	if tokenAddress != "" {
		if err := keys.ValidateEDDSAPublicKey(tokenAddress); err != nil {
			return types.ContractOutput{}, fmt.Errorf("invalid token address: %w", err)
		}
	}
	if programType != "" && programType != "fixed-percentage" && programType != "variable-percentage" {
		return types.ContractOutput{}, fmt.Errorf("invalid program_type: %s", programType)
	}
	if page < 1 {
		return types.ContractOutput{}, fmt.Errorf("page must be greater than 0")
	}
	if limit < 1 {
		return types.ContractOutput{}, fmt.Errorf("limit must be greater than 0")
	}

	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_LIST_CASHBACKS
	data := map[string]interface{}{
		"owner":         owner,
		"token_address": tokenAddress,
		"program_type":  programType,
		"paused":        paused,
		"page":          page,
		"limit":         limit,
		"ascending":     ascending,
	}

	return c.GetState(contractVersion, method, data)
}

func (c *networkClient) ClaimCashback(address, amount string) (types.ContractOutput, error) {
	if address == "" {
		return types.ContractOutput{}, fmt.Errorf("address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(address); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid address: %w", err)
	}
	if amount == "" {
		return types.ContractOutput{}, fmt.Errorf("amount not set")
	}

	from := c.publicKey
	if from == "" {
		return types.ContractOutput{}, fmt.Errorf("from address not set")
	}
	if err := keys.ValidateEDDSAPublicKey(from); err != nil {
		return types.ContractOutput{}, fmt.Errorf("invalid from address: %w", err)
	}

	to := address
	contractVersion := cashbackV1.CASHBACK_CONTRACT_V1
	method := cashbackV1.METHOD_CLAIM_CASHBACK

	data := map[string]interface{}{
		"address": address,
		"amount":  amount,
	}

	return c.SignAndSendTransaction(from, to, contractVersion, method, data)
}

File: dart/blockchain/lib/cashback.dart
part of 'two_finance_blockchain.dart';

extension CashbackClient on TwoFinanceBlockchain {
  static const String _programTypeFixed = 'fixed-percentage';
  static const String _programTypeVariable = 'variable-percentage';

  bool _isZeroLikeDateTime(DateTime value) {
    final isUnixEpoch = value.millisecondsSinceEpoch == 0;
    final isGoZeroLike =
        value.year <= 1 &&
        value.month == 1 &&
        value.day == 1 &&
        value.hour == 0 &&
        value.minute == 0 &&
        value.second == 0 &&
        value.millisecond == 0 &&
        value.microsecond == 0;

    return isUnixEpoch || isGoZeroLike;
  }

  String _requireFromAddress() {
    final from = publicKeyHex ?? '';
    if (from.isEmpty) {
      throw ArgumentError('from address not set');
    }
    KeyManager.validateEDDSAPublicKeyHex(from);
    return from;
  }

  void _validateProgramType(String programType) {
    if (programType != _programTypeFixed &&
        programType != _programTypeVariable) {
      throw ArgumentError('invalid program_type: $programType');
    }
  }

  Future<ContractOutput> addCashback({
    required String address,
    required String owner,
    required String tokenAddress,
    required String programType,
    required String percentage,
    required DateTime startAt,
    required DateTime expiredAt,
    required bool paused,
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (owner.isEmpty) throw ArgumentError('owner not set');
    if (tokenAddress.isEmpty) throw ArgumentError('token address not set');
    if (percentage.isEmpty) throw ArgumentError('percentage not set');
    if (_isZeroLikeDateTime(startAt)) {
      throw ArgumentError('start_at not set');
    }
    if (_isZeroLikeDateTime(expiredAt)) {
      throw ArgumentError('expired_at not set');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);
    KeyManager.validateEDDSAPublicKeyHex(owner);
    KeyManager.validateEDDSAPublicKeyHex(tokenAddress);
    _validateProgramType(programType);

    if (owner != from) {
      throw ArgumentError('owner must match from address');
    }

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_ADD_CASHBACK,
      data: {
        'address': address,
        'owner': owner,
        'token_address': tokenAddress,
        'program_type': programType,
        'percentage': percentage,
        'start_at': startAt.toUtc().toIso8601String(),
        'expired_at': expiredAt.toUtc().toIso8601String(),
        'paused': paused,
      },
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> updateCashback({
    required String address,
    String tokenAddress = '',
    required String programType,
    required String percentage,
    required DateTime startAt,
    required DateTime expiredAt,
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (percentage.isEmpty) throw ArgumentError('percentage not set');
    if (_isZeroLikeDateTime(startAt)) {
      throw ArgumentError('start_at not set');
    }
    if (_isZeroLikeDateTime(expiredAt)) {
      throw ArgumentError('expired_at not set');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);
    if (tokenAddress.isNotEmpty) {
      KeyManager.validateEDDSAPublicKeyHex(tokenAddress);
    }
    _validateProgramType(programType);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_UPDATE_CASHBACK,
      data: {
        'address': address,
        'token_address': tokenAddress,
        'program_type': programType,
        'percentage': percentage,
        'start_at': startAt.toUtc().toIso8601String(),
        'expired_at': expiredAt.toUtc().toIso8601String(),
      },
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> pauseCashback({
    required String address,
    required bool paused,
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (!paused) throw ArgumentError('paused must be true: Pause: $paused');
    KeyManager.validateEDDSAPublicKeyHex(address);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_PAUSE_CASHBACK,
      data: {'address': address, 'paused': paused},
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> unpauseCashback({
    required String address,
    required bool paused,
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (paused) throw ArgumentError('paused must be false: Pause: $paused');
    KeyManager.validateEDDSAPublicKeyHex(address);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_UNPAUSE_CASHBACK,
      data: {'address': address, 'paused': paused},
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> depositCashbackFunds({
    required String address,
    required String tokenAddress,
    required String amount,
    String tokenType = '',
    String uuid = '',
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (tokenAddress.isEmpty) throw ArgumentError('token address not set');
    if (amount.isEmpty) throw ArgumentError('amount not set');
    final effectiveTokenType = tokenType.isEmpty
        ? TOKEN_TYPE_FUNGIBLE
        : tokenType;
    if (effectiveTokenType == TOKEN_TYPE_NON_FUNGIBLE && uuid.isEmpty) {
      throw ArgumentError('uuid must be set for non-fungible tokens');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);
    KeyManager.validateEDDSAPublicKeyHex(tokenAddress);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_DEPOSIT_CASHBACK,
      data: {
        'address': address,
        'token_address': tokenAddress,
        'amount': amount,
        'token_type': effectiveTokenType,
        'uuid': uuid,
      },
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> withdrawCashbackFunds({
    required String address,
    required String tokenAddress,
    required String amount,
    String tokenType = '',
    String uuid = '',
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (tokenAddress.isEmpty) throw ArgumentError('token address not set');
    if (amount.isEmpty) throw ArgumentError('amount not set');
    final effectiveTokenType = tokenType.isEmpty
        ? TOKEN_TYPE_FUNGIBLE
        : tokenType;
    if (effectiveTokenType == TOKEN_TYPE_NON_FUNGIBLE && uuid.isEmpty) {
      throw ArgumentError('uuid must be set for non-fungible tokens');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);
    KeyManager.validateEDDSAPublicKeyHex(tokenAddress);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_WITHDRAW_CASHBACK,
      data: {
        'address': address,
        'token_address': tokenAddress,
        'amount': amount,
        'token_type': effectiveTokenType,
        'uuid': uuid,
      },
      version: 1,
      uuid7: newUUID7(),
    );
  }

  Future<ContractOutput> getCashback({required String address}) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('cashback address must be set');
    KeyManager.validateEDDSAPublicKeyHex(address);
    KeyManager.validateEDDSAPublicKeyHex(from);

    return getState(
      to: address,
      method: METHOD_GET_CASHBACK,
      data: const <String, dynamic>{},
    );
  }

  Future<ContractOutput> listCashbacks({
    String owner = '',
    String tokenAddress = '',
    String programType = '',
    required bool paused,
    int page = 1,
    int limit = 10,
    bool ascending = false,
  }) async {
    final from = _requireFromAddress();
    KeyManager.validateEDDSAPublicKeyHex(from);

    if (owner.isNotEmpty) {
      KeyManager.validateEDDSAPublicKeyHex(owner);
    }
    if (tokenAddress.isNotEmpty) {
      KeyManager.validateEDDSAPublicKeyHex(tokenAddress);
    }
    if (programType.isNotEmpty) {
      _validateProgramType(programType);
    }
    if (page < 1) throw ArgumentError('page must be greater than 0');
    if (limit < 1) throw ArgumentError('limit must be greater than 0');

    return getState(
      to: '',
      method: METHOD_LIST_CASHBACKS,
      data: {
        'owner': owner,
        'program_type': programType,
        'paused': paused,
        'page': page,
        'limit': limit,
        'ascending': ascending,
        'token_address': tokenAddress,
        'contract_version': CASHBACK_CONTRACT_V2,
      },
    );
  }

  Future<ContractOutput> claimCashback({
    required String address,
    required String amount,
    String tokenType = '',
    String uuid = '',
  }) async {
    final from = _requireFromAddress();

    if (address.isEmpty) throw ArgumentError('address not set');
    if (amount.isEmpty) throw ArgumentError('amount not set');
    final effectiveTokenType = tokenType.isEmpty
        ? TOKEN_TYPE_FUNGIBLE
        : tokenType;
    if (effectiveTokenType == TOKEN_TYPE_NON_FUNGIBLE && uuid.isEmpty) {
      throw ArgumentError('uuid must be set for non-fungible tokens');
    }

    KeyManager.validateEDDSAPublicKeyHex(address);

    return signAndSendTransaction(
      chainID: _chainID,
      from: from,
      to: address,
      method: METHOD_CLAIM_CASHBACK,
      data: {
        'address': address,
        'amount': amount,
        'token_type': effectiveTokenType,
        'uuid': uuid,
      },
      version: 1,
      uuid7: newUUID7(),
    );
  }
}
